scipy/sparse/lil.py

```diff
@@ -453,14 +453,23 @@ def tocsr(self, copy=False):
         lst = [len(x) for x in self.rows]
         idx_dtype = get_index_dtype(maxval=max(self.shape[1], sum(lst)))
         indptr = np.cumsum([0] + lst, dtype=idx_dtype)
-        indices = np.empty(indptr[-1], dtype=idx_dtype)
-        data = np.empty(indptr[-1], dtype=self.dtype)
-
-        start = 0
-        for stop, indices_i, data_i in zip(indptr[1:], self.rows, self.data):
-            indices[start:stop] = indices_i
-            data[start:stop] = data_i
-            start = stop
+        # choose faster construction approach depending on density 
+        if len(self.rows) == 0:
+            indices = np.empty(0, dtype=idx_dtype)
+            data = np.empty(0, dtype=self.dtype)            
+        elif indptr[-1] / len(self.rows) >= 25:
+            indices = np.empty(indptr[-1], dtype=idx_dtype)
+            data = np.empty(indptr[-1], dtype=self.dtype)
```

Comment:
Might be nice for readability to define `nnz = indptr[-1]`

Reply:
Yes, I updated the code.